File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] arr) {

        List<Integer> res = new ArrayList<Integer>();
        int minrow = 0;
        int mincol = 0;
        int maxrow = arr.length - 1;
        int maxcol = arr[0].length - 1;


        int tne = arr.length * arr[0].length;     //total numbers of elements
        int count = 0;

        while (count < tne) {
            //top wall
            if (count < tne) {
                for (int i = mincol; i <= maxcol; i++) {
                res.add(arr[minrow][i]);
                count++;
                }
            }
            minrow++;
            //right wall
            if (count < tne) {
                for (int i = minrow; i <= maxrow; i++) {
                res.add(arr[i][maxcol]);
                count++;
                }
            }
            maxcol--;
            //bottom wall
            if (count < tne) {
                for (int i = maxcol; i >= mincol; i--) {
                res.add(arr[maxrow][i]);
                count++;
                }
            }
            maxrow--;
            //left wall
            if (count < tne) {
                for (int i = maxrow; i >= minrow; i--) {
                res.add(arr[i][mincol]);
                count++;
                }
            }
            mincol++;
            
        }

        return res;
    }
}

File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& arr) {
        vector<int> ans;
        int maxC = arr[0].size() - 1;
        int maxR = arr.size() - 1, cnt=0;
        int minR = 0, minC = 0;
        int tne = (maxC+1)*(maxR+1);
        while(cnt < tne){
            //top
            if(cnt < tne){   
                for(int i=minC; i<=maxC; i++){
                    ans.push_back(arr[minR][i]);
                    cnt++;
                }
            }
            minR++;
            //right
            if(cnt < tne){
                for(int i=minR; i<=maxR; i++){
                    ans.push_back(arr[i][maxC]);
                    cnt++;
                }
            }
            maxC--;
            //bottom
            if(cnt < tne){
                for(int i=maxC; i>=minC; i--){
                    ans.push_back(arr[maxR][i]);
                    cnt++;
                }
            }
            maxR--;
            //left
            if(cnt < tne){
                for(int i=maxR; i>=minR; i--){
                    ans.push_back(arr[i][minC]);
                    cnt++;
                }
            }
            minC++;
        }
        return ans;
    }
};